pkg/agent/wireguard/client_test.go

```diff
@@ -0,0 +1,279 @@
+			for _, n := range c.AllowedIPs {
+				peer.AllowedIPs = append(peer.AllowedIPs, net.IPNet{
+					IP:   append(n.IP[:0:0], n.IP...),
+					Mask: append(n.Mask[:0:0], n.Mask...),
+				})
+			}
+			f.peers[c.PublicKey] = peer
+		}
+	}
+	return nil
+}
+func getFakeClient() *client {
+	return &client{
+		wgClient:       &fakeWireGuardClient{},
+		nodeName:       "fake-node-1",
+		mtu:            1420,
+		listenPort:     listenPort,
+		peerByNodeName: &sync.Map{},
+		enabledIPv4:    true,
+		enabledIPv6:    false,
+		nodeIPv4:       net.ParseIP("10.20.30.41"),
+	}
+}
+
+func assertWgPeerAndPeerConfigIdentical(t *testing.T, peer wgtypes.Peer, peercfg wgtypes.PeerConfig) {
+	assert.EqualValues(t, peer.AllowedIPs, peercfg.AllowedIPs)
+	assert.EqualValues(t, peer.Endpoint, peercfg.Endpoint)
+	assert.EqualValues(t, peer.PublicKey, peercfg.PublicKey)
+}
+
+func deepCopyForWireGuardPeer(peer wgtypes.Peer) wgtypes.Peer {
+	c := wgtypes.Peer{}
+	c.AllowedIPs = make([]net.IPNet, len(peer.AllowedIPs))
+	for idx := range peer.AllowedIPs {
+		ip := peer.AllowedIPs[idx].IP
+		mask := peer.AllowedIPs[idx].Mask
+		c.AllowedIPs[idx].IP = append(ip[:0:0], ip...)
+		c.AllowedIPs[idx].Mask = append(mask[:0:0], mask...)
+	}
+	if peer.Endpoint != nil {
+		ep := &net.UDPAddr{}
+		ip := peer.Endpoint.IP
+		ep.IP = append(ip[:0:0], ip...)
+		ep.Port = peer.Endpoint.Port
+		ep.Zone = peer.Endpoint.Zone
+		c.Endpoint = ep
+	}
+	c.PublicKey = peer.PublicKey
+	return c
+}
+
+func Test_RemoveStalePeers(t *testing.T) {
+	client := getFakeClient()
+	pk1, _ := wgtypes.GeneratePrivateKey()
+	pk2, _ := wgtypes.GeneratePrivateKey()
+	pk3, _ := wgtypes.GeneratePrivateKey()
+	t.Run("pass empty/nil slice should remove all existing peers", func(tt *testing.T) {
+		fc := &fakeWireGuardClient{
+			peers: map[wgtypes.Key]wgtypes.Peer{
+				pk1.PublicKey(): {PublicKey: pk1.PublicKey()},
+				pk2.PublicKey(): {PublicKey: pk2.PublicKey()},
+			},
+		}
+		client.wgClient = fc
+		err := client.RemoveStalePeers(nil)
+		assert.Nil(tt, err)
+		assert.Empty(tt, fc.peers)
+	})
+
+	t.Run("args has no intersection with existing peers", func(tt *testing.T) {
+		fc := &fakeWireGuardClient{
+			peers: map[wgtypes.Key]wgtypes.Peer{
+				pk1.PublicKey(): {PublicKey: pk1.PublicKey()},
+				pk2.PublicKey(): {PublicKey: pk2.PublicKey()},
+			},
+		}
+		client.wgClient = fc
+		err := client.RemoveStalePeers([]string{pk3.PublicKey().String()})
+		assert.Nil(tt, err)
+		assert.Empty(tt, fc.peers)
+	})
+
+	t.Run("should only keep peer in args", func(tt *testing.T) {
+		fc := &fakeWireGuardClient{
+			peers: map[wgtypes.Key]wgtypes.Peer{
+				pk1.PublicKey(): {PublicKey: pk1.PublicKey()},
+				pk2.PublicKey(): {PublicKey: pk2.PublicKey()},
+				pk3.PublicKey(): {PublicKey: pk3.PublicKey()},
+			},
+		}
+		client.wgClient = fc
+		err := client.RemoveStalePeers([]string{pk3.PublicKey().String()})
+		assert.Nil(tt, err)
+		assert.Len(tt, fc.peers, 1)
+		_, exist := fc.peers[pk3.PublicKey()]
+		assert.True(tt, exist)
```

Comment:
Maybe a table driven test is suitable here:
```
tests := []struct {
	name       string
	existingPeers   map[wgtypes.Key]wgtypes.Peer
	inputPublicKeys []string
	expectedPeers   map[wgtypes.Key]wgtypes.Peer
}{
}
for _, tt := range tests {
	t.Run(tt.name, func(t *testing.T) {
		client := getFakeClient()
		fc := &fakeWireGuardClient{peers: tt.existingPeers}
		client.wgClient = fc
		err := client.RemoveStalePeers(tt.inputPublicKeys)
		assert.Nil(t, err)
		assert.Equal(t, tt.expectedPeers, fc.peers)
	})
}
```

Reply:
done